src/runner_manager.py

```diff
@@ -0,0 +1,392 @@
+    ) -> RunnerApplication:
+        """Get the URL for the latest runner binary.
+
+        The runner binary URL changes when a new version is available.
+
+        Args:
+            os_name: Name of operating system.
+            arch_name: Name of architecture.
+
+        Returns:
+            Information on the runner application.
+        """
+        runner_bins: RunnerApplicationList = []
+        if isinstance(self.config.path, GitHubRepo):
+            runner_bins = self._clients.github.actions.list_runner_applications_for_repo(
+                owner=self.config.path.owner, repo=self.config.path.repo
+            )
+        if isinstance(self.config.path, GitHubOrg):
+            runner_bins = self._clients.github.actions.list_runner_application_for_org(
+                org=self.config.path.org
+            )
+
+        logger.debug("Response of runner binary list: %s", runner_bins)
+
+        try:
+            return next(
+                bin
+                for bin in runner_bins
+                if bin["os"] == os_name and bin["architecture"] == arch_name
+            )
+        except StopIteration as err:
+            raise RunnerBinaryError(
+                f"Unable query GitHub runner binary information for {os_name} {arch_name}"
+            ) from err
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def update_runner_bin(self, binary: RunnerApplication) -> None:
+        """Download a runner file, replacing the current copy.
+
+        Remove the existing runner binary to prevent it from being used. This
+        is done to prevent security issues arising from outdated runner binary
+        containing security flaws. The newest version of runner binary should
+        always be used.
+
+        Args:
+            binary: Information on the runner binary to download.
+        """
+        logger.info("Downloading runner binary from: %s", binary["download_url"])
+
+        # Delete old version of runner binary.
+        if self.runner_bin_path is not None:
+            self.runner_bin_path.unlink(missing_ok=True)
+            self.runner_bin_path = None
+
+        # Download the new file
+        response = self.session.get(binary["download_url"], stream=True)
+
+        logger.info(
+            "Download of runner binary from %s return status code: %i",
+            binary["download_url"],
+            response.status_code,
+        )
+
+        if not binary["sha256_checksum"]:
+            logger.error("Checksum for runner binary is not found, unable to verify download.")
+            raise RunnerBinaryError("Checksum for runner binary is not found in GitHub response.")
+
+        sha256 = hashlib.sha256()
+
+        # Use tempfile as recommend by bandit to avoid hardcoding the file
+        # path.  The `delete=False` arg save the file permanently for the charm
+        # to use start runners.
+        with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
```

Comment:
You can leave the context manager and use the `tempfile.NamedTemporaryFile` directly if you don't need to delete the temp file. This will reduce nesting layers.

`tmp_file = tempfile.NamedTemporaryFile(delete=False)`

Reply:
The context manager is for closing the file, else `tmp_file.close()` still needs to be called.